pkg/kubelet/kubeletconfig/util/files/files_test.go

```diff
@@ -0,0 +1,339 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package files
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+func TestFileExists(t *testing.T) {
+	fs := utilfs.DefaultFs{}
+	prePath := getTempDir(t, "TestFileExists")
+	defer cleanup(t, fs, prePath)
+
+	testcases := map[string]struct {
+		path         string // file path to be created
+		create       bool   // if should create the file before test
+		isFile       bool   // if the path indicates a file
+		expectedBool bool   // expected result in bool
+		expectedErr  error  // expected error
+	}{
+		"not exists": {
+			path:         prePath + "/NotExists",
+			create:       false,
+			isFile:       false,
+			expectedBool: false,
+			expectedErr:  nil,
+		},
+		"exists as dir": {
+			path:         prePath + "/ExistsAsDir",
+			create:       true,
+			isFile:       false,
+			expectedBool: false,
+			expectedErr:  fmt.Errorf("expected regular file at %q, but mode is %q", prePath + "/ExistsAsDir", "drwxr-xr-x"), // should have error, but the error info can only be fulfilled after creating the dir
+		},
+		"exists as file": {
+			path:         prePath + "/ExistsAsFile",
+			create:       true,
+			isFile:       true,
+			expectedBool: true,
+			expectedErr:  nil,
+		},
+	}
+
+	for tesetcaseName, testcase := range testcases {
+		t.Run(tesetcaseName, func(t *testing.T) {
+			if testcase.create {
+				if err := createFileOrDir(testcase.isFile, fs, testcase.path); err != nil {
+					t.Fatalf("unexpected error in creating %s, error: %v", testcase.path, err)
+				}
+			}
+
+			result, err := FileExists(fs, testcase.path)
+			if !reflect.DeepEqual(testcase.expectedErr, err) {
```

Comment:
you can just directly compare the error strings `testcase.expectedErr.Error() != err.Error()`

Reply:
This is my last nit, then lgtm.

I have some ideas in mind on how to abstract the testing of file interfaces more cleanly, but I have to prototype what I'm thinking before I'm sure it will work. In the meantime, I'd like to merge this since it gets us test coverage in an important place where we were previously missing it.